frontend/src/components/tabs/MetricsTab.test.tsx

```diff
@@ -0,0 +1,108 @@
+import React from 'react';
+import { QueryClient, QueryClientProvider } from 'react-query';
+import { BrowserRouter } from 'react-router-dom';
+import { MetricsTab } from './MetricsTab';
+
+describe('MetricsTab', () => {
+  const mlmdUtils = require('src/lib/MlmdUtils');
+  const OutputArtifactLoader = require('src/lib/OutputArtifactLoader');
+  const execution = new Execution();
+  const queryClient = new QueryClient();
+
+  beforeEach(() => {
+    const executionGetIdSpy = jest.spyOn(execution, 'getId');
+    executionGetIdSpy.mockReturnValue(123);
+  });
+
+  afterEach(() => {
+    jest.resetAllMocks();
+  });
+
+  it('shows ROC curve', async () => {
+    const artifact = new Artifact();
+    const artifactType = new ArtifactType();
+
+    const confidenceMetrics = {
+      list: [
+        {
+          confidenceThreshold: 2,
+          falsePositiveRate: 0,
+          recall: 0,
+        },
+        {
+          confidenceThreshold: 1,
+          falsePositiveRate: 0,
+          recall: 0.33962264150943394,
+        },
+        {
+          confidenceThreshold: 0.9,
+          falsePositiveRate: 0,
+          recall: 0.6037735849056604,
+        },
+      ],
+    };
+    artifact.getCustomPropertiesMap().set('name', new Value().setStringValue('metrics'));
+    artifact
+      .getCustomPropertiesMap()
+      .set(
+        'confidenceMetrics',
+        new Value().setStructValue(Struct.fromJavaScript(confidenceMetrics)),
+      );
+
+    jest
+      .spyOn(mlmdUtils, 'getOutputArtifactsInExecution')
+      .mockReturnValue(Promise.resolve([artifact]));
+    jest
+      .spyOn(OutputArtifactLoader, 'getArtifactTypes')
+      .mockReturnValue(Promise.resolve([artifactType]));
+    jest.spyOn(OutputArtifactLoader, 'filterArtifactsByType').mockReturnValue([artifact]);
+
+    const { getByText } = render(
+      <BrowserRouter>
+        <QueryClientProvider client={queryClient}>
+          <MetricsTab execution={execution}></MetricsTab>
+        </QueryClientProvider>
+      </BrowserRouter>,
+    );
+
+    getByText('Metrics is loading.');
+    // We should upgrade react-scripts for capability to use libraries normally:
+    // https://github.com/testing-library/dom-testing-library/issues/477
+    await waitFor(() => getByText('ROC Curve'));
+  });
+
+  it('has no metrics', async () => {
+    jest.spyOn(mlmdUtils, 'getOutputArtifactsInExecution').mockReturnValue(Promise.resolve([]));
+    jest.spyOn(OutputArtifactLoader, 'getArtifactTypes').mockReturnValue(Promise.resolve([]));
+    jest.spyOn(OutputArtifactLoader, 'filterArtifactsByType').mockReturnValue([]);
+
+    const { getByText } = render(
+      <BrowserRouter>
+        <QueryClientProvider client={queryClient}>
+          <MetricsTab execution={execution}></MetricsTab>
+        </QueryClientProvider>
+      </BrowserRouter>,
+    );
+
+    await waitFor(() => getByText('There is no metrics artifact available in this step.'));
+  });
+});
```

Comment:
nit: can we add one more test case when user written roc curve data spec is invalid?

We need to make sure UI doesn't crash and shows a helpful error message.

Reply:
I manually change the field name for type `ConfidenceMetric` to simulate a data spec mismatch. I verified that the UI doesn't crash, it is only that some information on ROC Curve is missing (For example: subset of `x, y, label` field is missing). How should I show helpful error message in this case? Since ROC Curve takes an array of `ConfidenceMetric`, it is possible that one of few data entry is missing, but the overall array provides enough data to render ROC Curve. I am leaning towards relaxing the restriction on data spec on this case.

Reply:
Added test to make sure we show error message when data spec is invalid.